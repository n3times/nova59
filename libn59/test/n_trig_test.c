#include "n_test.h"

#include <stdio.h>

#define PI 3.141592653590

typedef n_t (*trig_fun_t)(n_t, n_trig_t mode, n_err_t *err);

static void t(n_t n, n_trig_t mode, char *str, n_t (op)(n_t, n_trig_t, n_err_t *)) {
  n_err_t err;
  n_t res = op(n, mode, &err);
  char s[N_PRINT_MAX_SIZE];
  char s_res[N_PRINT_MAX_SIZE];
  printf("%4s %s => %s%s\n",
         str, n_print(n, s), n_print(res, s_res), err ? " ?" : "");
}

int main() {
  // Forensics.
  n_t N_9 = n_make(9000000000000LL, 0);
  n_trig_t D = N_DEG;
  n_t res = n_asin(n_acos(n_atan(n_tan(n_cos(n_sin(N_9, D, 0), D,
      0), D, 0), D, 0), D, 0), D, 0);
  char str[N_PRINT_MAX_SIZE];
  printf("Forensics:\n  asin(acos(atan(tan(cos(sin(9)))))) = %s\n\n\n",
         n_print(res, str));

  n_t n;
  trig_fun_t trig_funs[] = { n_sin, n_cos, n_tan, n_asin, n_acos, n_atan };
  char *trig_fun_strs[] = { "sin", "cos", "tan", "asin", "acos", "atan" };

  printf("DEGREES: -720 .. -90 -45 0 45 90 .. 720\n");
  printf("=======================================\n\n");
  for (int k = 0; k < N_ELEMS(trig_funs) / 2; k++) {
    for (int i = -16; i < 16; i += 1) {
      n = n_d2n(i * 45, NULL);
      t(n, N_DEG, trig_fun_strs[k], trig_funs[k]);
    }
    printf("\n");
  }
  printf("\n");

  printf("GRADIANS -800 .. -100 -50 0 50 100 .. 800\n");
  printf("=========================================\n\n");
  for (int k = 0; k < N_ELEMS(trig_funs) / 2; k++) {
    for (int i = -16; i < 16; i += 1) {
      n = n_d2n(i * 50, NULL);
      t(n, N_GRAD, trig_fun_strs[k], trig_funs[k]);
    }
    printf("\n");
  }
  printf("\n");

  printf("RADIANS -4PI .. -PI/2 -PI/4 0 PI/4 PI/2 .. 4PI\n");
  printf("==============================================\n\n");
  for (int k = 0; k < N_ELEMS(trig_funs) / 2; k++) {
    for (int i = -16; i < 16; i += 1) {
      n_t two = n_make(2000000000000LL, 0);
      n_t four = n_make(4000000000000LL, 0);
      n = n_times(n_div(N_PI, two, NULL), n_d2n(i/2, NULL), NULL);
      n = n_plus(n, n_times(n_div(N_PI, four, NULL), n_d2n(i%2, NULL), NULL), NULL);
      t(n, N_RAD, trig_fun_strs[k], trig_funs[k]);
    }
    printf("\n");
  }
  printf("\n");

  n_t ns[] = { TEST_NUMBERS };
  n_trig_t modes[] = { N_DEG, N_GRAD, N_RAD };
  char *mode_strs[] = { "Deg ", "Grad", "Rad "};
  for (int j = 0; j < N_ELEMS(modes); j++) {
    printf("%s: 0 1 pi eps inf -1 -pi -eps -inf\n", mode_strs[j]);
    printf("=====================================\n\n");
    for (int k = 0; k < N_ELEMS(trig_funs); k++) {
      for (int i = 0; i < N_ELEMS(ns); i++) {
        t(ns[i], modes[j], trig_fun_strs[k], trig_funs[k]);
      }
      printf("\n");
    }
    printf("\n");
  }

  printf("Values around pi and pi/2 RADIANS\n");
  printf("=================================\n\n");
  n_t ns2[] = {
                // Around pi.
                n_make(3141592000000LL, 0),
                n_make(3141592600000LL, 0),
                n_make(3141592653589LL, 0),
                n_make(3141592653590LL, 0),
                n_make(3141592653591LL, 0),
                n_make(3141592700000LL, 0),
                n_make(3141593000000LL, 0),

                // Around pi/2.
                n_make(1570796000000LL, 0),
                n_make(1570796300000LL, 0),
                n_make(1570796326794LL, 0),
                n_make(1570796326795LL, 0),
                n_make(1570796326796LL, 0),
                n_make(1570796400000LL, 0),
                n_make(1570797000000LL, 0),
              };
  for (int i = 0; i < N_ELEMS(trig_funs) / 2; i++) {
    for (int j = 0; j < N_ELEMS(ns2); j++) {
      t(ns2[j], N_RAD, trig_fun_strs[i], trig_funs[i]);
    }
    printf("\n");
  }
  printf("\n");
}
