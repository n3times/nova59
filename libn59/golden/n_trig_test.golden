Forensics:
  asin(acos(atan(tan(cos(sin(9)))))) =  8999999981719 00


DEGREES: -720 .. -90 -45 0 45 90 .. 720
=======================================

 sin -7200000000000 02 =>  0000000000000 00
 sin -6750000000000 02 =>  7071067811860-01
 sin -6300000000000 02 =>  1000000000000 00
 sin -5850000000000 02 =>  7071067811870-01
 sin -5400000000000 02 =>  0000000000000 00
 sin -4950000000000 02 => -7071067811861-01
 sin -4500000000000 02 => -1000000000000 00
 sin -4050000000000 02 => -7071067811869-01
 sin -3600000000000 02 =>  0000000000000 00
 sin -3150000000000 02 =>  7071067811863-01
 sin -2700000000000 02 =>  1000000000000 00
 sin -2250000000000 02 =>  7071067811867-01
 sin -1800000000000 02 =>  0000000000000 00
 sin -1350000000000 02 => -7071067811864-01
 sin -9000000000000 01 => -1000000000000 00
 sin -4500000000000 01 => -7071067811866-01
 sin  0000000000000 00 =>  0000000000000 00
 sin  4500000000000 01 =>  7071067811866-01
 sin  9000000000000 01 =>  1000000000000 00
 sin  1350000000000 02 =>  7071067811864-01
 sin  1800000000000 02 =>  0000000000000 00
 sin  2250000000000 02 => -7071067811867-01
 sin  2700000000000 02 => -1000000000000 00
 sin  3150000000000 02 => -7071067811863-01
 sin  3600000000000 02 =>  0000000000000 00
 sin  4050000000000 02 =>  7071067811869-01
 sin  4500000000000 02 =>  1000000000000 00
 sin  4950000000000 02 =>  7071067811861-01
 sin  5400000000000 02 =>  0000000000000 00
 sin  5850000000000 02 => -7071067811870-01
 sin  6300000000000 02 => -1000000000000 00
 sin  6750000000000 02 => -7071067811860-01

 cos -7200000000000 02 =>  1000000000000 00
 cos -6750000000000 02 =>  7071067811871-01
 cos -6300000000000 02 =>  0000000000000 00
 cos -5850000000000 02 => -7071067811861-01
 cos -5400000000000 02 => -1000000000000 00
 cos -4950000000000 02 => -7071067811869-01
 cos -4500000000000 02 =>  0000000000000 00
 cos -4050000000000 02 =>  7071067811862-01
 cos -3600000000000 02 =>  1000000000000 00
 cos -3150000000000 02 =>  7071067811868-01
 cos -2700000000000 02 =>  0000000000000 00
 cos -2250000000000 02 => -7071067811864-01
 cos -1800000000000 02 => -1000000000000 00
 cos -1350000000000 02 => -7071067811867-01
 cos -9000000000000 01 =>  0000000000000 00
 cos -4500000000000 01 =>  7071067811865-01
 cos  0000000000000 00 =>  1000000000000 00
 cos  4500000000000 01 =>  7071067811865-01
 cos  9000000000000 01 =>  0000000000000 00
 cos  1350000000000 02 => -7071067811867-01
 cos  1800000000000 02 => -1000000000000 00
 cos  2250000000000 02 => -7071067811864-01
 cos  2700000000000 02 =>  0000000000000 00
 cos  3150000000000 02 =>  7071067811868-01
 cos  3600000000000 02 =>  1000000000000 00
 cos  4050000000000 02 =>  7071067811862-01
 cos  4500000000000 02 =>  0000000000000 00
 cos  4950000000000 02 => -7071067811869-01
 cos  5400000000000 02 => -1000000000000 00
 cos  5850000000000 02 => -7071067811861-01
 cos  6300000000000 02 =>  0000000000000 00
 cos  6750000000000 02 =>  7071067811871-01

 tan -7200000000000 02 =>  0000000000000 00
 tan -6750000000000 02 =>  9999999999985-01
 tan -6300000000000 02 =>  9999999999999 99 ?
 tan -5850000000000 02 => -1000000000001 00
 tan -5400000000000 02 =>  0000000000000 00
 tan -4950000000000 02 =>  9999999999989-01
 tan -4500000000000 02 =>  9999999999999 99 ?
 tan -4050000000000 02 => -1000000000001 00
 tan -3600000000000 02 =>  0000000000000 00
 tan -3150000000000 02 =>  9999999999993-01
 tan -2700000000000 02 =>  9999999999999 99 ?
 tan -2250000000000 02 => -1000000000001 00
 tan -1800000000000 02 =>  0000000000000 00
 tan -1350000000000 02 =>  9999999999997-01
 tan -9000000000000 01 =>  9999999999999 99 ?
 tan -4500000000000 01 => -1000000000000 00
 tan  0000000000000 00 =>  0000000000000 00
 tan  4500000000000 01 =>  1000000000000 00
 tan  9000000000000 01 =>  9999999999999 99 ?
 tan  1350000000000 02 => -9999999999997-01
 tan  1800000000000 02 =>  0000000000000 00
 tan  2250000000000 02 =>  1000000000001 00
 tan  2700000000000 02 =>  9999999999999 99 ?
 tan  3150000000000 02 => -9999999999993-01
 tan  3600000000000 02 =>  0000000000000 00
 tan  4050000000000 02 =>  1000000000001 00
 tan  4500000000000 02 =>  9999999999999 99 ?
 tan  4950000000000 02 => -9999999999989-01
 tan  5400000000000 02 =>  0000000000000 00
 tan  5850000000000 02 =>  1000000000001 00
 tan  6300000000000 02 =>  9999999999999 99 ?
 tan  6750000000000 02 => -9999999999985-01


GRADIANS -800 .. -100 -50 0 50 100 .. 800
=========================================

 sin -8000000000000 02 =>  0000000000000 00
 sin -7500000000000 02 =>  7071067811860-01
 sin -7000000000000 02 =>  1000000000000 00
 sin -6500000000000 02 =>  7071067811870-01
 sin -6000000000000 02 =>  0000000000000 00
 sin -5500000000000 02 => -7071067811861-01
 sin -5000000000000 02 => -1000000000000 00
 sin -4500000000000 02 => -7071067811869-01
 sin -4000000000000 02 =>  0000000000000 00
 sin -3500000000000 02 =>  7071067811863-01
 sin -3000000000000 02 =>  1000000000000 00
 sin -2500000000000 02 =>  7071067811867-01
 sin -2000000000000 02 =>  0000000000000 00
 sin -1500000000000 02 => -7071067811864-01
 sin -1000000000000 02 => -1000000000000 00
 sin -5000000000000 01 => -7071067811866-01
 sin  0000000000000 00 =>  0000000000000 00
 sin  5000000000000 01 =>  7071067811866-01
 sin  1000000000000 02 =>  1000000000000 00
 sin  1500000000000 02 =>  7071067811864-01
 sin  2000000000000 02 =>  0000000000000 00
 sin  2500000000000 02 => -7071067811867-01
 sin  3000000000000 02 => -1000000000000 00
 sin  3500000000000 02 => -7071067811863-01
 sin  4000000000000 02 =>  0000000000000 00
 sin  4500000000000 02 =>  7071067811869-01
 sin  5000000000000 02 =>  1000000000000 00
 sin  5500000000000 02 =>  7071067811861-01
 sin  6000000000000 02 =>  0000000000000 00
 sin  6500000000000 02 => -7071067811870-01
 sin  7000000000000 02 => -1000000000000 00
 sin  7500000000000 02 => -7071067811860-01

 cos -8000000000000 02 =>  1000000000000 00
 cos -7500000000000 02 =>  7071067811871-01
 cos -7000000000000 02 =>  0000000000000 00
 cos -6500000000000 02 => -7071067811861-01
 cos -6000000000000 02 => -1000000000000 00
 cos -5500000000000 02 => -7071067811869-01
 cos -5000000000000 02 =>  0000000000000 00
 cos -4500000000000 02 =>  7071067811862-01
 cos -4000000000000 02 =>  1000000000000 00
 cos -3500000000000 02 =>  7071067811868-01
 cos -3000000000000 02 =>  0000000000000 00
 cos -2500000000000 02 => -7071067811864-01
 cos -2000000000000 02 => -1000000000000 00
 cos -1500000000000 02 => -7071067811867-01
 cos -1000000000000 02 =>  0000000000000 00
 cos -5000000000000 01 =>  7071067811865-01
 cos  0000000000000 00 =>  1000000000000 00
 cos  5000000000000 01 =>  7071067811865-01
 cos  1000000000000 02 =>  0000000000000 00
 cos  1500000000000 02 => -7071067811867-01
 cos  2000000000000 02 => -1000000000000 00
 cos  2500000000000 02 => -7071067811864-01
 cos  3000000000000 02 =>  0000000000000 00
 cos  3500000000000 02 =>  7071067811868-01
 cos  4000000000000 02 =>  1000000000000 00
 cos  4500000000000 02 =>  7071067811862-01
 cos  5000000000000 02 =>  0000000000000 00
 cos  5500000000000 02 => -7071067811869-01
 cos  6000000000000 02 => -1000000000000 00
 cos  6500000000000 02 => -7071067811861-01
 cos  7000000000000 02 =>  0000000000000 00
 cos  7500000000000 02 =>  7071067811871-01

 tan -8000000000000 02 =>  0000000000000 00
 tan -7500000000000 02 =>  9999999999985-01
 tan -7000000000000 02 =>  9999999999999 99 ?
 tan -6500000000000 02 => -1000000000001 00
 tan -6000000000000 02 =>  0000000000000 00
 tan -5500000000000 02 =>  9999999999989-01
 tan -5000000000000 02 =>  9999999999999 99 ?
 tan -4500000000000 02 => -1000000000001 00
 tan -4000000000000 02 =>  0000000000000 00
 tan -3500000000000 02 =>  9999999999993-01
 tan -3000000000000 02 =>  9999999999999 99 ?
 tan -2500000000000 02 => -1000000000001 00
 tan -2000000000000 02 =>  0000000000000 00
 tan -1500000000000 02 =>  9999999999997-01
 tan -1000000000000 02 =>  9999999999999 99 ?
 tan -5000000000000 01 => -1000000000000 00
 tan  0000000000000 00 =>  0000000000000 00
 tan  5000000000000 01 =>  1000000000000 00
 tan  1000000000000 02 =>  9999999999999 99 ?
 tan  1500000000000 02 => -9999999999997-01
 tan  2000000000000 02 =>  0000000000000 00
 tan  2500000000000 02 =>  1000000000001 00
 tan  3000000000000 02 =>  9999999999999 99 ?
 tan  3500000000000 02 => -9999999999993-01
 tan  4000000000000 02 =>  0000000000000 00
 tan  4500000000000 02 =>  1000000000001 00
 tan  5000000000000 02 =>  9999999999999 99 ?
 tan  5500000000000 02 => -9999999999989-01
 tan  6000000000000 02 =>  0000000000000 00
 tan  6500000000000 02 =>  1000000000001 00
 tan  7000000000000 02 =>  9999999999999 99 ?
 tan  7500000000000 02 => -9999999999985-01


RADIANS -4PI .. -PI/2 -PI/4 0 PI/4 PI/2 .. 4PI
==============================================

 sin -1256637061436 01 =>  0000000000000 00
 sin -1178097245095 01 =>  7071067811948-01
 sin -1099557428756 01 =>  1000000000000 00
 sin -1021017612416 01 =>  7071067811817-01
 sin -9424777960770 00 =>  0000000000000 00
 sin -8639379797372 00 => -7071067811865-01
 sin -7853981633975 00 => -1000000000000 00
 sin -7068583470577 00 => -7071067811865-01
 sin -6283185307180 00 =>  0000000000000 00
 sin -5497787143782 00 =>  7071067811866-01
 sin -4712388980385 00 =>  1000000000000 00
 sin -3926990816987 00 =>  7071067811864-01
 sin -3141592653590 00 =>  0000000000000 00
 sin -2356194490192 00 => -7071067811868-01
 sin -1570796326795 00 => -1000000000000 00
 sin -7853981633975-01 => -7071067811866-01
 sin  0000000000000 00 =>  0000000000000 00
 sin  7853981633975-01 =>  7071067811866-01
 sin  1570796326795 00 =>  1000000000000 00
 sin  2356194490192 00 =>  7071067811868-01
 sin  3141592653590 00 =>  0000000000000 00
 sin  3926990816987 00 => -7071067811864-01
 sin  4712388980385 00 => -1000000000000 00
 sin  5497787143782 00 => -7071067811866-01
 sin  6283185307180 00 =>  0000000000000 00
 sin  7068583470577 00 =>  7071067811865-01
 sin  7853981633975 00 =>  1000000000000 00
 sin  8639379797372 00 =>  7071067811865-01
 sin  9424777960770 00 =>  0000000000000 00
 sin  1021017612416 01 => -7071067811817-01
 sin  1099557428756 01 => -1000000000000 00
 sin  1178097245095 01 => -7071067811948-01

 cos -1256637061436 01 =>  1000000000000 00
 cos -1178097245095 01 =>  7071067811783-01
 cos -1099557428756 01 =>  0000000000000 00
 cos -1021017612416 01 => -7071067811914-01
 cos -9424777960770 00 => -1000000000000 00
 cos -8639379797372 00 => -7071067811866-01
 cos -7853981633975 00 =>  0000000000000 00
 cos -7068583470577 00 =>  7071067811866-01
 cos -6283185307180 00 =>  1000000000000 00
 cos -5497787143782 00 =>  7071067811865-01
 cos -4712388980385 00 =>  0000000000000 00
 cos -3926990816987 00 => -7071067811867-01
 cos -3141592653590 00 => -1000000000000 00
 cos -2356194490192 00 => -7071067811863-01
 cos -1570796326795 00 =>  0000000000000 00
 cos -7853981633975-01 =>  7071067811865-01
 cos  0000000000000 00 =>  1000000000000 00
 cos  7853981633975-01 =>  7071067811865-01
 cos  1570796326795 00 =>  0000000000000 00
 cos  2356194490192 00 => -7071067811863-01
 cos  3141592653590 00 => -1000000000000 00
 cos  3926990816987 00 => -7071067811867-01
 cos  4712388980385 00 =>  0000000000000 00
 cos  5497787143782 00 =>  7071067811865-01
 cos  6283185307180 00 =>  1000000000000 00
 cos  7068583470577 00 =>  7071067811866-01
 cos  7853981633975 00 =>  0000000000000 00
 cos  8639379797372 00 => -7071067811866-01
 cos  9424777960770 00 => -1000000000000 00
 cos  1021017612416 01 => -7071067811914-01
 cos  1099557428756 01 =>  0000000000000 00
 cos  1178097245095 01 =>  7071067811783-01

 tan -1256637061436 01 =>  0000000000000 00
 tan -1178097245095 01 =>  1000000000023 00
 tan -1099557428756 01 =>  9999999999999 99 ?
 tan -1021017612416 01 => -9999999999863-01
 tan -9424777960770 00 =>  0000000000000 00
 tan -8639379797372 00 =>  9999999999999-01
 tan -7853981633975 00 =>  9999999999999 99 ?
 tan -7068583470577 00 => -9999999999999-01
 tan -6283185307180 00 =>  0000000000000 00
 tan -5497787143782 00 =>  1000000000000 00
 tan -4712388980385 00 =>  9999999999999 99 ?
 tan -3926990816987 00 => -9999999999995-01
 tan -3141592653590 00 =>  0000000000000 00
 tan -2356194490192 00 =>  1000000000001 00
 tan -1570796326795 00 =>  9999999999999 99 ?
 tan -7853981633975-01 => -1000000000000 00
 tan  0000000000000 00 =>  0000000000000 00
 tan  7853981633975-01 =>  1000000000000 00
 tan  1570796326795 00 =>  9999999999999 99 ?
 tan  2356194490192 00 => -1000000000001 00
 tan  3141592653590 00 =>  0000000000000 00
 tan  3926990816987 00 =>  9999999999995-01
 tan  4712388980385 00 =>  9999999999999 99 ?
 tan  5497787143782 00 => -1000000000000 00
 tan  6283185307180 00 =>  0000000000000 00
 tan  7068583470577 00 =>  9999999999999-01
 tan  7853981633975 00 =>  9999999999999 99 ?
 tan  8639379797372 00 => -9999999999999-01
 tan  9424777960770 00 =>  0000000000000 00
 tan  1021017612416 01 =>  9999999999863-01
 tan  1099557428756 01 =>  9999999999999 99 ?
 tan  1178097245095 01 => -1000000000023 00


Deg : 0 1 pi eps inf -1 -pi -eps -inf
=====================================

 sin  0000000000000 00 =>  0000000000000 00
 sin  1000000000000 00 =>  1745240643728-02
 sin  3141592653590 00 =>  5480366514880-02
 sin  1000000000000-99 =>  1000000000000-99 ?
 sin  9999999999999 99 =>  0000000000000 00
 sin -1000000000000 00 => -1745240643728-02
 sin -3141592653590 00 => -5480366514880-02
 sin -1000000000000-99 => -1000000000000-99 ?
 sin -9999999999999 99 =>  0000000000000 00

 cos  0000000000000 00 =>  1000000000000 00
 cos  1000000000000 00 =>  9998476951564-01
 cos  3141592653590 00 =>  9984971498639-01
 cos  1000000000000-99 =>  1000000000000 00
 cos  9999999999999 99 =>  1000000000000 00
 cos -1000000000000 00 =>  9998476951564-01
 cos -3141592653590 00 =>  9984971498639-01
 cos -1000000000000-99 =>  1000000000000 00
 cos -9999999999999 99 =>  1000000000000 00

 tan  0000000000000 00 =>  0000000000000 00
 tan  1000000000000 00 =>  1745506492822-02
 tan  3141592653590 00 =>  5488615080801-02
 tan  1000000000000-99 =>  1000000000000-99 ?
 tan  9999999999999 99 =>  0000000000000 00
 tan -1000000000000 00 => -1745506492822-02
 tan -3141592653590 00 => -5488615080801-02
 tan -1000000000000-99 => -1000000000000-99 ?
 tan -9999999999999 99 =>  0000000000000 00

asin  0000000000000 00 =>  0000000000000 00
asin  1000000000000 00 =>  8999999999999 01
asin  3141592653590 00 =>  3141592653590 00 ?
asin  1000000000000-99 =>  5729577951308-98
asin  9999999999999 99 =>  9999999999999 99 ?
asin -1000000000000 00 => -8999999999999 01
asin -3141592653590 00 => -3141592653590 00 ?
asin -1000000000000-99 => -5729577951308-98
asin -9999999999999 99 => -9999999999999 99 ?

acos  0000000000000 00 =>  8999999999999 01
acos  1000000000000 00 =>  0000000000000 00
acos  3141592653590 00 =>  3141592653590 00 ?
acos  1000000000000-99 =>  8999999999999 01
acos  9999999999999 99 =>  9999999999999 99 ?
acos -1000000000000 00 =>  1800000000000 02
acos -3141592653590 00 => -3141592653590 00 ?
acos -1000000000000-99 =>  8999999999999 01
acos -9999999999999 99 => -9999999999999 99 ?

atan  0000000000000 00 =>  0000000000000 00
atan  1000000000000 00 =>  4500000000000 01
atan  3141592653590 00 =>  7234321284858 01
atan  1000000000000-99 =>  5729577951308-98
atan  9999999999999 99 =>  8999999999999 01
atan -1000000000000 00 => -4500000000000 01
atan -3141592653590 00 => -7234321284858 01
atan -1000000000000-99 => -5729577951308-98
atan -9999999999999 99 => -8999999999999 01


Grad: 0 1 pi eps inf -1 -pi -eps -inf
=====================================

 sin  0000000000000 00 =>  0000000000000 00
 sin  1000000000000 00 =>  1570731731182-02
 sin  3141592653590 00 =>  4932799550253-02
 sin  1000000000000-99 =>  1000000000000-99 ?
 sin  9999999999999 99 =>  0000000000000 00
 sin -1000000000000 00 => -1570731731182-02
 sin -3141592653590 00 => -4932799550253-02
 sin -1000000000000-99 => -1000000000000-99 ?
 sin -9999999999999 99 =>  0000000000000 00

 cos  0000000000000 00 =>  1000000000000 00
 cos  1000000000000 00 =>  9998766324817-01
 cos  3141592653590 00 =>  9987826334392-01
 cos  1000000000000-99 =>  1000000000000 00
 cos  9999999999999 99 =>  1000000000000 00
 cos -1000000000000 00 =>  9998766324817-01
 cos -3141592653590 00 =>  9987826334392-01
 cos -1000000000000-99 =>  1000000000000 00
 cos -9999999999999 99 =>  1000000000000 00

 tan  0000000000000 00 =>  0000000000000 00
 tan  1000000000000 00 =>  1570925532367-02
 tan  3141592653590 00 =>  4938811894704-02
 tan  1000000000000-99 =>  1000000000000-99 ?
 tan  9999999999999 99 =>  0000000000000 00
 tan -1000000000000 00 => -1570925532367-02
 tan -3141592653590 00 => -4938811894704-02
 tan -1000000000000-99 => -1000000000000-99 ?
 tan -9999999999999 99 =>  0000000000000 00

asin  0000000000000 00 =>  0000000000000 00
asin  1000000000000 00 =>  9999999999999 01
asin  3141592653590 00 =>  3141592653590 00 ?
asin  1000000000000-99 =>  6366197723675-98
asin  9999999999999 99 =>  9999999999999 99 ?
asin -1000000000000 00 => -9999999999999 01
asin -3141592653590 00 => -3141592653590 00 ?
asin -1000000000000-99 => -6366197723675-98
asin -9999999999999 99 => -9999999999999 99 ?

acos  0000000000000 00 =>  9999999999999 01
acos  1000000000000 00 =>  0000000000000 00
acos  3141592653590 00 =>  3141592653590 00 ?
acos  1000000000000-99 =>  9999999999999 01
acos  9999999999999 99 =>  9999999999999 99 ?
acos -1000000000000 00 =>  2000000000000 02
acos -3141592653590 00 => -3141592653590 00 ?
acos -1000000000000-99 =>  9999999999999 01
acos -9999999999999 99 => -9999999999999 99 ?

atan  0000000000000 00 =>  0000000000000 00
atan  1000000000000 00 =>  5000000000000 01
atan  3141592653590 00 =>  8038134760954 01
atan  1000000000000-99 =>  6366197723675-98
atan  9999999999999 99 =>  9999999999999 01
atan -1000000000000 00 => -5000000000000 01
atan -3141592653590 00 => -8038134760954 01
atan -1000000000000-99 => -6366197723675-98
atan -9999999999999 99 => -9999999999999 01


Rad : 0 1 pi eps inf -1 -pi -eps -inf
=====================================

 sin  0000000000000 00 =>  0000000000000 00
 sin  1000000000000 00 =>  8414709848079-01
 sin  3141592653590 00 =>  0000000000000 00
 sin  1000000000000-99 =>  1000000000000-99
 sin  9999999999999 99 =>  0000000000000 00
 sin -1000000000000 00 => -8414709848079-01
 sin -3141592653590 00 =>  0000000000000 00
 sin -1000000000000-99 => -1000000000000-99
 sin -9999999999999 99 =>  0000000000000 00

 cos  0000000000000 00 =>  1000000000000 00
 cos  1000000000000 00 =>  5403023058681-01
 cos  3141592653590 00 => -1000000000000 00
 cos  1000000000000-99 =>  1000000000000 00
 cos  9999999999999 99 =>  1000000000000 00
 cos -1000000000000 00 =>  5403023058681-01
 cos -3141592653590 00 => -1000000000000 00
 cos -1000000000000-99 =>  1000000000000 00
 cos -9999999999999 99 =>  1000000000000 00

 tan  0000000000000 00 =>  0000000000000 00
 tan  1000000000000 00 =>  1557407724655 00
 tan  3141592653590 00 =>  0000000000000 00
 tan  1000000000000-99 =>  1000000000000-99
 tan  9999999999999 99 =>  0000000000000 00
 tan -1000000000000 00 => -1557407724655 00
 tan -3141592653590 00 =>  0000000000000 00
 tan -1000000000000-99 => -1000000000000-99
 tan -9999999999999 99 =>  0000000000000 00

asin  0000000000000 00 =>  0000000000000 00
asin  1000000000000 00 =>  1570796326795 00
asin  3141592653590 00 =>  3141592653590 00 ?
asin  1000000000000-99 =>  1000000000000-99
asin  9999999999999 99 =>  9999999999999 99 ?
asin -1000000000000 00 => -1570796326795 00
asin -3141592653590 00 => -3141592653590 00 ?
asin -1000000000000-99 => -1000000000000-99
asin -9999999999999 99 => -9999999999999 99 ?

acos  0000000000000 00 =>  1570796326795 00
acos  1000000000000 00 =>  0000000000000 00
acos  3141592653590 00 =>  3141592653590 00 ?
acos  1000000000000-99 =>  1570796326795 00
acos  9999999999999 99 =>  9999999999999 99 ?
acos -1000000000000 00 =>  3141592653590 00
acos -3141592653590 00 => -3141592653590 00 ?
acos -1000000000000-99 =>  1570796326795 00
acos -9999999999999 99 => -9999999999999 99 ?

atan  0000000000000 00 =>  0000000000000 00
atan  1000000000000 00 =>  7853981633974-01
atan  3141592653590 00 =>  1262627255679 00
atan  1000000000000-99 =>  1000000000000-99
atan  9999999999999 99 =>  1570796326795 00
atan -1000000000000 00 => -7853981633974-01
atan -3141592653590 00 => -1262627255679 00
atan -1000000000000-99 => -1000000000000-99
atan -9999999999999 99 => -1570796326795 00


Values around pi and pi/2 RADIANS
=================================

 sin  3141592000000 00 =>  6535897935203-07
 sin  3141592600000 00 =>  5358979317006-08
 sin  3141592653589 00 =>  7932657934721-13
 sin  3141592653590 00 =>  0000000000000 00
 sin  3141592653591 00 => -1206912007693-12
 sin  3141592700000 00 => -4641020666628-08
 sin  3141593000000 00 => -3464102066194-07
 sin  1570796000000 00 =>  9999999999999-01
 sin  1570796300000 00 =>  1000000000000 00
 sin  1570796326794 00 =>  1000000000000 00
 sin  1570796326795 00 =>  1000000000000 00
 sin  1570796326796 00 =>  1000000000000 00
 sin  1570796400000 00 =>  1000000000000 00
 sin  1570797000000 00 =>  9999999999998-01

 cos  3141592000000 00 => -9999999999998-01
 cos  3141592600000 00 => -1000000000000 00
 cos  3141592653589 00 => -1000000000000 00
 cos  3141592653590 00 => -1000000000000 00
 cos  3141592653591 00 => -1000000000000 00
 cos  3141592700000 00 => -1000000000000 00
 cos  3141593000000 00 => -9999999999999-01
 cos  1570796000000 00 =>  3267948967602-07
 cos  1570796300000 00 =>  2679489658503-08
 cos  1570796326794 00 =>  8966773470272-13
 cos  1570796326795 00 =>  0000000000000 00
 cos  1570796326796 00 => -1103278409533-12
 cos  1570796400000 00 => -7320510325131-08
 cos  1570797000000 00 => -6732051033795-07

 tan  3141592000000 00 => -6535897935205-07
 tan  3141592600000 00 => -5358979317006-08
 tan  3141592653589 00 => -7932657934721-13
 tan  3141592653590 00 =>  0000000000000 00
 tan  3141592653591 00 =>  1206912007693-12
 tan  3141592700000 00 =>  4641020666628-08
 tan  3141593000000 00 =>  3464102066194-07
 tan  1570796000000 00 =>  3060023304874 06
 tan  1570796300000 00 =>  3732053963435 07
 tan  1570796326794 00 =>  1115228351999 12
 tan  1570796326795 00 =>  9999999999999 99 ?
 tan  1570796326796 00 => -9063895308381 11
 tan  1570796400000 00 => -1366024984033 07
 tan  1570797000000 00 => -1485431401188 06


